README.md
# XDigger is a game for XWindows.
You are a little (digger-)man and have to collect diamonds.

-------------------------------------------------------------------------------

install:

 * xutils-dev libx11-dev libxext-dev
 * make
 * gcc
 * libfuse2 osspd osspd-pulseaudio (/dev/dsp <- osspd.service)

Now you can test the game. If you want to install the game
do the following two steps:

 * xmkmf
 * sed -i 's/-D_BSD_SOURCE -D_SVID_SOURCE/-D_DEFAULT_SOURCE/g' Makefile (imake bug workaround)
 * make
 * ./xdigger -h
 * ./xdigger -6 -sound dsp
 * make install (as root)
 * make install.man (as root)

If you want to use NAS (Network Audio System) or RPlay, edit first the 
configure.h file.

-------------------------------------------------------------------------------

Levels created by:
 1 Heiko Schade
 2 Heiko Schade
 3 Heiko Schade
 4 Martin Guth
 5 Martin Guth
 6 Martin Guth
 7
 8
 9
10
11
12
13 ?2
14
15 Heiko Schade
16 Thomas Carstens
17 Alexander Lang
18
19 Alexander Lang
20
21 Alexander Lang
22 Martin Guth
23
24 Alexander Lang
25 Alexander Lang
26 Martin Guth
27 Martin Guth
28
29 Alexander Lang
30 Alexander Lang


Original programmed by:
Alexander Lang
lang@mueritz-comp.de



ChangeLog
22/11/07
  - fixed a lot of compiler warnings

99/07/04
  - fixed bug with 32bit-displaydepth

98/05/04
  - changed some colors

98/04/22
  - some changes in event-handling

97/12/21
  - added dynamic change of size
  - option -noblink turns off diamond-blinking

97/02/17
  - NAS (network audio system) supported

96/11/07
  - diamonds now are blinking

96/06/21
  - fixed problems with SUN - compability

96/06/17
  - added experimental rplay - support

96/06/04
  - added KC85-sound emulation for /dev/audio & /dev/dsp (linux) by Jan Fricke


sound.h
/* XDigger  Copyright (C) 1988-99 Alexander Lang.

XDigger is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

XDigger is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; see the file COPYING.  If not, write to
the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */

#ifndef _SOUND_H

#define _SOUND_H

#define TON_SCHRITT 0
#define TON_STEINE 1
#define TON_DIAMANT 2

extern void sound_init();
extern void sound(char ton_typ);
extern void sound_done();
extern void Fill_TonBuffer(int ton_low, int ton_high, int korrektur, Bool sndheader);
extern int sound_device;
extern int ton_laenge[3];
extern char ton_buffer[3][2000];

#define SD_AUTO -1
#define SD_NONE  0
#define SD_DSP   1
#define SD_AUDIO 2
#define SD_XBELL 3
#define SD_RPLAY 4
#define SD_NAS   5

#define CPU_FREQ_KC85_2   1750000   //CPU-Takt KC85/2
#define CPU_FREQ_KC85_3   1750000   //CPU-Takt KC85/3
#define CPU_FREQ_KC85_4   1773447.5 //CPU-Takt KC85/4
#define KC_CTC_FREQ_VT16  109375    //CPU_FREQ_KC85_3 / 16  //VT16  -> 109375Hz
#define KC_CTC_FREQ_VT256 6836      //CPU_FREQ_KC85_3 / 256 //VT256 -> 6835,9375Hz

#define TON_DSP_LOW 0x40
#define TON_DSP_HIGH 0xc0
#define TON_DSP_RATE 44100
#define TON_AUDIO_LOW 0x10
#define TON_AUDIO_HIGH 0x70
#define TON_AUDIO_RATE 8000
#define TON_NAS_LOW 0x10
#define TON_NAS_HIGH 0x70

#endif /* _SOUND_H */


title.c
/* XDigger  Copyright (C) 1988-99 Alexander Lang.

XDigger is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

XDigger is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; see the file COPYING.  If not, write to
the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */

#include <stdio.h>
#include <string.h>
#include <X11/keysym.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/XKBlib.h>

#include "drawpixmaps.h"
#include "drawtext.h"
#include "icon.h"
#include "xdigger.h"

void Init_Title()
{
  int i;

/*   for (i=0; i<32; i++) */
/*     WriteTextStr("                                        ",  */
/* 		 0, i, kcf_weiss, kcb_blau); */
  WriteTextStr("\234\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\236", 0, 0, kcf_weiss, kcb_blau);
  WriteTextStr("\237\240\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\241\242\237", 0, 1, kcf_weiss, kcb_blau);
  WriteTextStr("\237\243\234\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\236\251\237", 0, 2, kcf_weiss, kcb_blau);

  for (i=3; i<27+2*vertvar; i++)
  {
    WriteTextStr("\237\243\237", 0, i, kcf_weiss, kcb_blau);
    WriteTextStr("\237\251\237", 37, i, kcf_weiss, kcb_blau);
  }

  WriteTextStr("\200\201\202\203 \210 \211\212\213\214 \211\212\213\214 \200\201\215\215 \133\215\221\222", 7, 6, kcf_weiss, kcb_blau);
  WriteTextStr("\133 \206\207 \133 \133  \215 \133  \215 \133    \133 \223\224", 7, 7, kcf_weiss, kcb_blau);
  WriteTextStr("\133  \133 \133 \133 \244\216 \133 \244\216 \133\217\220  \133\225\226\227", 7, 8, kcf_weiss, kcb_blau);
  WriteTextStr("\133 \260\261 \133 \133 \316\133 \133 \316\133 \133    \133 \230\231", 7, 9, kcf_weiss, kcb_blau);
  WriteTextStr("\252\253\254\255 \262 \263\264\265\266 \263\264\265\266 \252\253\267\267 \133 \232\233", 7, 10, kcf_weiss, kcb_blau);

  WriteTextStr("WRITTEN BY  ALEXANDER LANG", 7, 13, kcf_weiss, kcb_blau);
/*   WriteTextStr("GRAPHIX BY  MARTIN    GUTH", 7, 15, kcf_weiss, kcb_blau); */
  WriteTextStr("GRAPHIX BY  STEFAN  DAHLKE", 7, 15, kcf_weiss, kcb_blau);
  WriteTextStr("HUMBOLDT-UNIVERSITY     \245\246", 7, 17, kcf_weiss, kcb_blau);
  WriteTextStr("         BERLIN         \247\250", 7, 18, kcf_weiss, kcb_blau);
  WriteTextStr("P: PLAY", 11, 20, kcf_weiss, kcb_blau);
  WriteTextStr("L: A LOOK AT THE ROOMS", 9, 22, kcf_weiss, kcb_blau);
  WriteTextStr("\140 1988", 26, 25, kcf_weiss, kcb_blau);
  WriteTextStr("BERLIN", 26, 26, kcf_weiss, kcb_blau);

  WriteTextStr("\237\243\306\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\310\251\237", 0, 27+2*vertvar, kcf_weiss, kcb_blau);
  WriteTextStr("\237\312\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\313\314\237", 0, 28+2*vertvar, kcf_weiss, kcb_blau);
  WriteTextStr("\306\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\235\310", 0, 29+2*vertvar, kcf_weiss, kcb_blau);
} /* Init_Title() */

/*void ClearExposeEvents()
{
  XEvent event;
  
  while (XCheckWindowEvent(display, mainwindow, ExposureMask, &event));
}*/

char Title()
{
  XSetWindowAttributes setwindowattributes;
  XWindowAttributes windowattributes;
  unsigned long valuemask;
  XEvent event;
  Bool Done = False;
  KeySym keysym;

  strcpy(LastArgv, " (Title)");
  setwindowattributes.background_pixel = kcb_blau;
  valuemask = CWBackPixel;
  XChangeWindowAttributes(display, mainwindow, valuemask, &setwindowattributes);

  ClearCharField(kcb_blau, True);
  Init_Title();
  DrawTextField();

  while (!Done)
    {
      if (!XEventsQueued(display, QueuedAfterReading))
	{
	  XDigger_Idle();
	  XFlush(display);
	} /* if (!XEventsQueued()) */
      else
	{
	  XNextEvent(display, &event);
          if (event.xany.window == iconwindow)
            Handle_IconWindow_Event(&event); else
	  switch(event.type)
	    {
	    case Expose:
              /*fprintf(stderr, "%d %d %d %d \n", event.xexpose.x, event.xexpose.y,
                      event.xexpose.width, event.xexpose.height);
              fflush(stderr);*/
              /* Resize ? */
              XGetWindowAttributes(display, mainwindow, &windowattributes);
              if (xpmmag != windowattributes.width / 320)
                ResizeXPixmaps(windowattributes.width / 320);
	      Mark_CharField(&event.xexpose, 29+2*vertvar);
	      if (event.xexpose.count == 0)
		DrawTextField();
	      break;
              /*case ResizeRequest:*/
            case MappingNotify:
	      XRefreshKeyboardMapping(&event.xmapping);
	      break;
	    case ButtonPress:
	      if (event.xbutton.button == Button1) return('p');
	      break;
	    case KeyPress:
	      /* keysym = XKeycodeToKeysym(display, event.xkey.keycode, 0); */
	      keysym = XkbKeycodeToKeysym(display, event.xkey.keycode, 0, event.xkey.state & ShiftMask ? 1 : 0);
	      switch (keysym)
              {
		case XK_e: return('e');
		case XK_l: return('l');
		case XK_p: return('p');
                case XK_q: 
		  quit = True;
		  return('q');
                case XK_plus:
                case XK_KP_Add:
                  XResizeWindow(display, mainwindow, 320*(xpmmag+1),
                                (240 + vertvar*16)*(xpmmag+1));
                  break;
                case XK_minus:
                case XK_KP_Subtract:
                  if (xpmmag > 1)
                    XResizeWindow(display, mainwindow, 320*(xpmmag-1),
                                  (240 + vertvar*16)*(xpmmag-1));
                  break;
              }
	      break;
	    case ClientMessage:
	      if (event.xclient.data.l[0] == protocol[0])
		{
		  /*exit(0);*/
		  quit = True;
                  return('q');                  
		}
	      break;
            /*case NoExpose:
              break;
            default:
              fprintf(stderr, "default: %d\n", event.type);
              fflush(stderr);*/
	    } /* switch(event.type) */
	} /* if (!XEventsQueued()) else */
    } /* while (!Done) */
  strcpy(LastArgv, "");
  return(0);
} /* char Title() */


drawtext.c
/* XDigger  Copyright (C) 1988-99 Alexander Lang.

XDigger is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

XDigger is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; see the file COPYING.  If not, write to
the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */

#include <stdio.h>
#include <string.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>

#include "typedef.h"
#include "xdigger.h"
#include "drawpixmaps.h"

CharEntry charfield[40][32];
GC gc_text;
XGCValues gc_text_values;
unsigned long gc_text_mask;

void ClearCharField(int bg, Bool todraw)
{
  int x, y;
    for (y=0; y<32; y++) for (x=0; x<40; x++)
      {
	charfield[x][y].c = '\0';
	charfield[x][y].bg = bg;
	charfield[x][y].todraw = todraw;
      }
} /* ClearCharField(int bg, Bool todraw) */

void DrawBitmapChar(GC gc, unsigned char c, int x, int y)
{
  if (c<0x20) c=0x20;
/*   XCopyPlane(display, xbm_charset, mainwindow, gc, 0,  */
/* 	     (c-0x20)*8*xpmmag, 8*xpmmag, 8*xpmmag, x*8*xpmmag, y*8*xpmmag, 1); */
  XCopyPlane(display, xbm_charset[c-0x20], mainwindow, gc,
	     0, 0, 8*xpmmag, 8*xpmmag, x*8*xpmmag, y*8*xpmmag, 1);
} /* DrawBitmapChar(GC gc, unsigned char c, int x, int y) */

void DrawOneCharEntry(int x, int y)
{
  if ((gc_text_values.foreground != charfield[x][y].fg) ||
      (gc_text_values.background != charfield[x][y].bg))
    {
      gc_text_values.foreground = charfield[x][y].fg;
      gc_text_values.background = charfield[x][y].bg;
      gc_text_mask = GCForeground | GCBackground;
      XChangeGC(display, gc_text, gc_text_mask, &gc_text_values);
    }
  DrawBitmapChar(gc_text, charfield[x][y].c, x, y);
} /* DrawOneCharEntry(int x, int y) */

void DrawTextField()
{
  int x, y;

  for (y=0; y<32; y++) for (x=0; x<40; x++)
    if (charfield[x][y].todraw)
      {
	DrawOneCharEntry(x, y);
	charfield[x][y].todraw = False;
      }
/*   XFlush(display); */
} /* DrawTextField() */

void WriteTextStr(const char *str, int x, int y, int fg, int bg)
{
  int i, len;

  len = strlen(str);
  for (i=0; i<len; i++)
    {
      charfield[x+i][y].c      = str[i];
      charfield[x+i][y].fg     = fg;
      charfield[x+i][y].bg     = bg;
      charfield[x+i][y].todraw = True;
    }
} /* WriteTextStr(const unsigned char *str, int x, int y, int fg, int bg) */

void Mark_CharField(XExposeEvent *xexpose, int max_y)
{
  int x, xa, xe, y, ya, ye;

  /*fprintf(stderr, "Mark %d \n", xpmmag);
  fflush(stderr);*/
  xa = (xexpose->x / (8*xpmmag));
  xe = ((xexpose->x + xexpose->width - 1) / (8*xpmmag));
  if (xa > 39) return;
  if (xe > 39) xe = 39;
  ya = (xexpose->y / (8*xpmmag));
  ye = ((xexpose->y + xexpose->height - 1) / (8*xpmmag));
  if (ya > max_y) return;
  if (ye > max_y) ye = max_y;
  for (y=ya; y<=ye; y++) for (x=xa; x<=xe; x++)
    charfield[x][y].todraw = True;
} /* Mark_CharField(XExposeEvent *xexpose, int max_y) */


drawtext.h
/* XDigger  Copyright (C) 1988-99 Alexander Lang.

XDigger is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

XDigger is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; see the file COPYING.  If not, write to
the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */

#ifndef _DRAWTEXT_H

#define _DRAWTEXT_H

#include <X11/Xlib.h>
#include "typedef.h"

extern CharEntry charfield[40][32];
extern GC gc_text;
extern XGCValues gc_text_values;
extern unsigned long gc_text_mask;

extern void ClearCharField(int bg, Bool todraw);
extern void DrawTextField();
extern void WriteTextStr(char *str, int x, int y, int fg, int bg);
extern void Mark_CharField(XExposeEvent *xexpose, int max_y);

#endif


sound.c
/* XDigger  Copyright (C) 1988-99 Alexander Lang.

XDigger is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

XDigger is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; see the file COPYING.  If not, write to
the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */

#define USE_PIPE

#include <errno.h>
#include <fcntl.h>
#include <netdb.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/ipc.h>
#ifdef USE_PIPE
#else
#include <sys/msg.h>
#endif
#include <sys/time.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#include <X11/Xlib.h>

#include "configure.h"

#ifdef linux
#include <linux/soundcard.h>
#endif
#ifdef SOUND_RPLAY
#include <rplay.h>
#endif
#ifdef SOUND_NAS
#include <audio/audiolib.h>
#include <audio/soundlib.h>
#endif
#include "xdigger.h"
#include "sound.h"

int ton_laenge[3];
char ton_buffer[3][2000];
int sound_device = SD_AUTO;
Bool soundserver_started = False;

#ifdef USE_PIPE
int filedes[2];
#else
int msgqid;
#endif

#ifdef SOUND_NAS
AuServer *auserver = NULL;
char *aureturn_status;
AuBucketID aubucketids[3];
#endif

void StartSoundServer()
{
  int fork_ok;
  
#ifdef USE_PIPE
  if(pipe(filedes) == -1)
    {
      fprintf(stderr, "%s: No more pipes.\n", progname);
      sound_device = SD_XBELL;
      return;
    }
#else
  struct msqid_ds buf;
  
  msgqid = msgget(IPC_PRIVATE, IPC_CREAT);
  if(msgqid == -1)
    {
      fprintf(stderr, "%s: No more private keys.\n", progname);
      sound_device = SD_XBELL;
      return;
    }
  printf("Got Key %d.\n", msgqid);
  
  msgctl(msgqid, IPC_STAT, &buf);
  buf.msg_perm.mode = 0600;
  msgctl(msgqid, IPC_SET, &buf);
#endif
  
  fork_ok = fork();
  if (fork_ok == -1)
    {
      fprintf(stderr, "%s: fork failed.\n", progname);
      sound_device = SD_XBELL;
#ifdef USE_PIPE
#else
      msgctl(msgqid, IPC_RMID, NULL);
#endif
      return;
    }
  
  if (fork_ok == 0)
    {
      
      /* Das ist der Soundserver */
      
      Bool done = False;
      int fd, i;
#ifdef USE_PIPE
#else
      struct msgbuf msgp;
#endif
      
      for (i = 1; i < pargc; i++) {strcpy(pargv[i], "");}
      strcat(pargv[0], " (Soundserver)");
      if (sound_device == SD_DSP) strcat(pargv[0], " (dsp)");
      if (sound_device == SD_AUDIO) strcat(pargv[0], " (audio)");
      
      /*       strcpy(LastArgv, " (Soundserver)"); */

      while (!done)
	{
	  char ton_typ;
	  
#ifdef USE_PIPE
 	  read(filedes[0], &ton_typ ,1); 
	  if((ton_typ>=0) && (ton_typ<3))
	    {
#else
	  if(msgrcv(msgqid, &msgp, 1, 0, 0) == -1)
	    {
	      /* if (errno != ENOMSG) */
/* 	      perror(progname); */
	      fprintf(stderr, "%s: (msgrcv) %s on id %d.\n",
		      progname, strerror(errno), msgqid);
	      continue;
	    }
	  
	  if(msgp.mtype == 2)
	    {
	      ton_typ = msgp.mtext[0];
#endif
#ifdef linux
	      if (sound_device == SD_DSP)
		{
		  int rate = TON_DSP_RATE;
		  int channels = 1;
		  int bits = 8;
	      
		  fd = open("/dev/dsp", O_WRONLY, 0); 
		  ioctl(fd, SOUND_PCM_WRITE_RATE, &rate);
      ioctl(fd, SOUND_PCM_WRITE_CHANNELS, &channels);
		  ioctl(fd, SOUND_PCM_WRITE_BITS, &bits);
		}
#endif
	      if (sound_device == SD_AUDIO)
		{
		  fd = open("/dev/audio", O_WRONLY, 0);
		}

	      write(fd, ton_buffer[(int)ton_typ], ton_laenge[(int)ton_typ]);
	      close(fd);
	    }
	  else
	    done = True;
	}
#ifdef linux
      if(sound_device == SD_DSP)
	ioctl(fd, SNDCTL_DSP_RESET, 0);
#endif
      exit(0);
    }
  else
    {
      soundserver_started = True;
    }
} /* StartSoundServer */

void Create_SND_Header(char *header, unsigned long data_size)
{
  int sample_rate = TON_AUDIO_RATE;
  
  strcpy(header, ".snd");
  header[4] = 0; header[5] = 0; header[6] = 0; header[7] = 24;
  header[8] = data_size >> 24; header[9] = data_size >> 16;
  header[10] = data_size >> 8; header[11] = data_size;
  header[12] = 0; header[13] = 0; header[14] = 0; header[15] = 1;
  header[16] = sample_rate >> 24; header[17] = sample_rate >> 16;
  header[18] = sample_rate >> 8; header[19] = sample_rate;
  header[20] = 0; header[21] = 0; header[22] = 0; header[23] = 1;
}

void Fill_TonBuffer(int ton_low, int ton_high, int rate, Bool sndheader)
{

  int i, j, k, n, offset;
  unsigned char augenblicklicherpeak;

  if (sndheader) offset = 24; else offset = 0;

  /* TON_SCHRITT */
  augenblicklicherpeak=ton_low;
  for (i=0,j=2; j>0; j--) {
    for(k=0; k<0x40; k++) {
      for (n=0+offset; n <= ( rate / KC_CTC_FREQ_VT256 <<0 ); n++)
        ton_buffer[TON_SCHRITT][ ( rate * i / KC_CTC_FREQ_VT256 <<0 ) + n ] = augenblicklicherpeak;
      i++;
    }
    augenblicklicherpeak = ton_low + ton_high - augenblicklicherpeak;
  }
  ton_laenge[TON_SCHRITT] = (rate * (i-1) / KC_CTC_FREQ_VT256 <<0) + n-1;

  /* TON_STEINE */
  augenblicklicherpeak=ton_low;
  for (i=0,j=0x0ff; ; j++) {
    if (j > 256) j = 1;
    if (j == 0x12) break; //j^TC-Reihe: 255,256,1...18 (^0x14 loops)
    for(k=0;k<j;k++){
      for (n = 0; n <= ( rate / KC_CTC_FREQ_VT16 <<0 ); n++)
        ton_buffer[TON_STEINE][ ( rate * i / KC_CTC_FREQ_VT16 <<0 ) + n ] = augenblicklicherpeak;
      i++;
    }
    augenblicklicherpeak = ton_low + ton_high - augenblicklicherpeak;
  }
  ton_laenge[TON_STEINE] = (rate * (i-1) / KC_CTC_FREQ_VT16 <<0) + n-1;

  /* TON_DIAMANT */
  augenblicklicherpeak=ton_low;
  for (i=0,j=0x40; j>0; j--) {
    for(k=0;k<j;k++) {
      for (n = 0; n <= ( rate / KC_CTC_FREQ_VT16 <<0 ); n++)
        ton_buffer[TON_DIAMANT][ ( rate * i / KC_CTC_FREQ_VT16 <<0 ) + n ] = augenblicklicherpeak;
      i++;
    }
    augenblicklicherpeak = ton_low + ton_high - augenblicklicherpeak;
  }
  ton_laenge[TON_DIAMANT] = (rate * (i-1) / KC_CTC_FREQ_VT16 <<0) + n-1;

  if (sndheader)
    {
      Create_SND_Header(ton_buffer[TON_SCHRITT], ton_laenge[TON_SCHRITT]);
      Create_SND_Header(ton_buffer[TON_STEINE], ton_laenge[TON_STEINE]);
      Create_SND_Header(ton_buffer[TON_DIAMANT], ton_laenge[TON_DIAMANT]);
    }
  
} /* Fill_TonBuffer(int ton_low, int ton_high, int korrektur, Bool sndheader)*/

#ifdef linux
Bool Check_DSP(Bool msg)
{
  int fd;
  int rate = TON_DSP_RATE;
  int channels = 1;
  int bits = 8;

  fd = open("/dev/dsp", O_WRONLY, 0);
  if ((fd == -1) ||
      (ioctl(fd, SOUND_PCM_WRITE_RATE, &rate) == -1) ||
      (ioctl(fd, SOUND_PCM_WRITE_CHANNELS, &channels) == -1) ||
      (ioctl(fd, SOUND_PCM_WRITE_BITS, &bits) == -1))
  {
    if (fd != -1) close(fd);
    if (msg || debug)
      fprintf(stderr, "%s: couldn't initialize /dev/dsp\n", progname);
    return(False);
  }
  close(fd);
  sound_device = SD_DSP;
  if (debug) fprintf(stderr, "%s: sound is dsp\n", progname);
  Fill_TonBuffer(TON_DSP_LOW, TON_DSP_HIGH, TON_DSP_RATE, False);
  StartSoundServer();
  return(True);
} /* Check_DSP */
#endif

Bool Check_SUN_Audio(Bool msg)
{
  int fd;

  fd = open("/dev/audio", O_WRONLY, 0);
  if(fd == -1)
  {
    if (msg || debug)
      fprintf(stderr, "%s: couldn't initialize /dev/audio\n", progname);
    return(False);
  }
  close(fd);
  sound_device = SD_AUDIO;
  if (debug) fprintf(stderr, "%s: sound is audio\n", progname);
  Fill_TonBuffer(TON_AUDIO_LOW, TON_AUDIO_HIGH, TON_AUDIO_RATE, False);
  StartSoundServer();
  return(True);
} /* Check_SUN_Audio */

#ifdef SOUND_NAS
Bool Check_NAS(Bool msg)
{
  int i, sample_rate;
  
  auserver = AuOpenServer(display_name, 0, NULL, 0, NULL, NULL);
  if (auserver == NULL)
  {
    if (msg || debug)
      fprintf(stderr, "%s: couldn't initialize NAS\n", progname);
    return(False);
  }
  else
  {
    sample_rate = AuServerMaxSampleRate(auserver);
    if (sample_rate > TON_DSP_RATE) sample_rate = TON_DSP_RATE;
    Fill_TonBuffer(TON_NAS_LOW, TON_NAS_HIGH, sample_rate, False);
    for (i=0; i<3; i++)
    {
      aubucketids[i] = 
        AuSoundCreateBucketFromData(auserver,
                                    SoundCreate(SoundFileFormatSnd,
                                                AuFormatLinearUnsigned8, 1,
                                                sample_rate,
                                                ton_laenge[i], NULL),
                                    ton_buffer[i],
                                    0, NULL, NULL);
      if (aubucketids[i] == AuNone)
      {
        if (msg || debug)
          fprintf(stderr, "%s: AuSoundCreateBucketFromData failed", progname);
        AuCloseServer(auserver);
        return(False);
      }
    }
    sound_device = SD_NAS;
    if (debug) fprintf(stderr, "%s: sound is NAS\n", progname);
    return(True);
  }
}
#endif

Bool Check_XBell(Bool msg)
{
  sound_device = SD_XBELL;
  if (debug) fprintf(stderr, "%s: sound is xbell\n", progname);
  return(True);
}

Bool XDisplay_is_on_Localhost()
{
  char localhost[1024], xhost[1024];
  char *c;
  /*struct hostent localhost_ent, xhost_ent;*/
    
  gethostname(localhost, sizeof(localhost));
  strcpy(xhost, DisplayString(display));
  c = strchr(xhost, ':');
  if (c) *c = 0; else xhost[0] = 0;
  if (strlen(xhost) == 0) return(True);

  strcpy(localhost, gethostbyname(localhost)->h_name);
  strcpy(xhost, gethostbyname(xhost)->h_name);
  if (debug)
    fprintf(stderr, "%s: localhost=%s\n             xhost=%s\n",
            progname, localhost, xhost);
  return(strcmp(localhost, xhost) == 0);
}

void sound_init()
{
  if (sound_device == SD_NONE) return;

  if (sound_device == SD_AUTO)
  {
    if (XDisplay_is_on_Localhost())
    {
      if (debug) fprintf(stderr, "%s: XDisplay is on localhost\n", progname);
#ifdef SOUND_DSP_AUDIO
#ifdef linux
      Check_DSP(False);
#endif
      if (sound_device == SD_AUTO) Check_SUN_Audio(False);
#endif
    }
    else
    {
      if (debug) fprintf(stderr, "%s: XDisplay is not on localhost\n",
                         progname);
    }
#ifdef SOUND_NAS
    if (sound_device == SD_AUTO) Check_NAS(False);
#endif
#ifdef SOUND_RPLAY
#endif
    if (sound_device == SD_AUTO) Check_XBell(False);
    if (sound_device == SD_AUTO)
    {
      sound_device = SD_NONE;
      return;
    }
  }
  else
  {
#ifdef SOUND_DSP_AUDIO
#ifdef linux
    if (sound_device == SD_DSP)
    {
      if (!Check_DSP(True)) sound_device = SD_NONE;
    }
#endif
    if (sound_device == SD_AUDIO)
    {
      if (!Check_SUN_Audio(True)) sound_device = SD_NONE;
    }
#endif
#ifdef SOUND_NAS
    if (sound_device == SD_NAS)
    {
      if (!Check_NAS(True)) sound_device = SD_NONE;
    }
#endif
#ifdef SOUND_RPLAY
#endif
    if (sound_device == SD_XBELL)
    {
      if (!Check_XBell(True)) sound_device = SD_NONE;
    }
  }
}

#ifdef SOUND_RPLAY
int Play_RPlay_Sound(char *soundfile, int volume)
{
  int rplay_fd;
  RPLAY *rp;
  
  rplay_fd = rplay_open_display(); if (rplay_fd == -1) return(-1);
  rp = rplay_create(RPLAY_PLAY); if (rp == NULL) return(-1);

  rplay_set(rp, RPLAY_APPEND, RPLAY_SOUND, soundfile, RPLAY_VOLUME, 
	    volume, RPLAY_PRIORITY, 255, NULL);
  rplay(rplay_fd, rp);

  rplay_close(rplay_fd);
  rplay_destroy(rp);
  return(0);
}
#endif

#ifdef SOUND_NAS
AuFlowID lastflow = -1;

void Play_NAS_Sound(char ton_typ)
{
  if (auserver != NULL)
  {
    if ((0 <= ton_typ) && (ton_typ <= 2))
    {
      /*AuDestroyFlow(auserver, lastflow, NULL);*/
      if (AuSoundPlayFromBucket(auserver, aubucketids[(int)ton_typ], AuNone,
                                AuFixedPointFromSum(1,0),
                                NULL, NULL, 1,
                                &lastflow, NULL, NULL, NULL) == NULL)
        fprintf(stderr, "%s: AuSoundPlayFromBucket failed", progname);
    }
    AuFlush(auserver);
  }
}
#endif

void sound(char ton_typ)
{
#ifdef SOUND_DSP_AUDIO

  if (soundserver_started) 
    {
#ifdef USE_PIPE
      write(filedes[1],&ton_typ,1);
#else
      struct msgbuf msgp;
      
      msgp.mtype = 2;
      msgp.mtext[0] = ton_typ;
      msgsnd(msgqid, &msgp, 1, 0);
#endif
    }
  if (sound_device == SD_XBELL) 
    if (ton_typ == TON_DIAMANT) XBell(display, -50);

#endif

#ifdef SOUND_RPLAY
  if (sound_device == SD_RPLAY)
    {
      char name[256] = XDIGGER_LIB_DIR;
      char error[20];

      switch (ton_typ)
	{
	case TON_DIAMANT:
	  strcat(name, "/diamond.au");
	  break;
	case TON_SCHRITT:
	  strcat(name, "/step.au");
	  break;
	case TON_STEINE:
	  strcat(name, "/stone.au");
	  break;
	}
      
/*       if (rplay_host_volume(hostname, name, 255) < 0) */
/*       if (rplay_display(name) < 0) */
      if (Play_RPlay_Sound(name, 200) < 0)
	{
	  sprintf(error, "%s: (rplay) ", progname);
	  rplay_perror(error);
	  fprintf(stderr, "%s: disable rplay-sound.\n", progname);
	  sound_device = SD_NONE;
	}

    }
#endif

#ifdef SOUND_NAS
  if (sound_device == SD_NAS)
  {
    Play_NAS_Sound(ton_typ);
  }
#endif
}

void sound_done()
{
#ifdef SOUND_DSP_AUDIO

  if (soundserver_started)
    {
#ifdef USE_PIPE
      char beenden = -1;

      write(filedes[1], &beenden, 1);
      wait(0);
#else
      struct msgbuf msgp;

      msgp.mtype = 1;
      msgsnd(msgqid, &msgp, 1, 0);
      wait(0);
      msgctl(msgqid, IPC_RMID, NULL);
#endif
    }
#endif

#ifdef SOUND_RPLAY
#endif

#ifdef SOUND_NAS
  if (sound_device == SD_NAS)
  {
    if (auserver != NULL)
    {
      int i;
      
      for (i=0; i<3; i++)
        AuDestroyBucket(auserver, aubucketids[i], NULL);
      AuCloseServer(auserver);
    }
  }
#endif
} /* sound_done */
